### PR TITLE
added jwt-based sign-in with role. and permission classes

### PR DESCRIPTION
**Pull request by ne0ascorbinka**

### CODE REVIEW

File: forum/users/permissions.py
Comment (mehalyna) [inline forum/users/permissions.py:1]: Directly accessing `request.auth.get('role')` assumes that:
- `request.auth` is a dictionary.
- `request.auth` always exists and contains the role key.
If authentication fails or `request.auth` is `None`, this will throw an `AttributeError`.

Instead of relying on `request.auth`, you should safely access the role field from the authenticated user object (`request.user`). Ensure that the user is authenticated first.
```
from rest_framework.permissions import BasePermission


class Unassigned(BasePermission):
    def has_permission(self, request, view):
        # Ensure the user is authenticated and has the role attribute
        return request.user.is_authenticated and getattr(request.user, 'role', None) == 'unassigned'


class IsInvestor(BasePermission):
    def has_permission(self, request, view):
        return request.user.is_authenticated and getattr(request.user, 'role', None) == 'investor'


class IsStartup(BasePermission):
    def has_permission(self, request, view):
        return request.user.is_authenticated and getattr(request.user, 'role', None) == 'startup'
```

Comment (mehalyna) [inline forum/users/permissions.py:1]: Hardcoding role values like 'investor' or 'startup' can lead to inconsistencies and errors. Instead, use constants or choices defined in the model.
Our CustomUser model has role choices as follows:
```
ROLE_CHOICES = [
    (0, "unassigned"),
    (1, "startup"),
    (2, "investor"),
]
```
Define role constants:
```
class UserRoles:
    UNASSIGNED = 0
    STARTUP = 1
    INVESTOR = 2
```
Then update the permissions:
```
from rest_framework.permissions import BasePermission
from .models import UserRoles


class Unassigned(BasePermission):
    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.role == UserRoles.UNASSIGNED


class IsInvestor(BasePermission):
    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.role == UserRoles.INVESTOR


class IsStartup(BasePermission):
    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.role == UserRoles.STARTUP
```

Comment (mehalyna) [inline forum/users/permissions.py:1]: Instead of creating multiple classes for each role, you can create a single generic permission class that takes a role as input.
```
from rest_framework.permissions import BasePermission

class HasRole(BasePermission):
    def __init__(self, role):
        self.role = role

    def has_permission(self, request, view):
        return request.user.is_authenticated and getattr(request.user, 'role', None) == self.role
```
And usage in views:
```
from .permissions import HasRole
from .models import UserRoles

class SomeView(APIView):
    permission_classes = [HasRole(UserRoles.INVESTOR)]
```

Comment (mehalyna) [inline forum/users/permissions.py:1]: Consider test these permissions in the following scenarios:
- User is not authenticated.
- `request.user` does not have a role attribute.
- User with the correct role (e.g., investor, startup) gains access.
- User with an incorrect role gets denied.

File: forum/users/serializers.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from rest_framework_simplejwt.serializers import TokenObtainPairSerializer
+
+class CustomTokenObtainPairSerializer(TokenObtainPairSerializer):
+    @classmethod
+    def get_token(cls, user, role='unassigned'):
```
Comment (mehalyna) [inline forum/users/serializers.py:5]: Issues:
- `role='unassigned'`:
  - This hardcodes the role to 'unassigned' unless explicitly passed.
  - It completely ignores the actual role field from the user model, which defeats the purpose of assigning dynamic roles.
- Not pulling the role from the user object:
  - The role should reflect the actual value stored in the database for the authenticated user.

File: forum/users/serializers.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+        token = super().get_token(user)
+        token['role'] = role
+        return token
```
Comment (mehalyna) [inline forum/users/serializers.py:8]: The `get_token` method should fetch the role from the user instance and add it to the token.
```
from rest_framework_simplejwt.serializers import TokenObtainPairSerializer

class CustomTokenObtainPairSerializer(TokenObtainPairSerializer):
    @classmethod
    def get_token(cls, user):
        # Get the token using the base class
        token = super().get_token(user)

        # Add custom claims (e.g., user role)
        token['role'] = getattr(user, 'role', 'unassigned')

        return token
```

File: forum/users/serializers.py
Comment (mehalyna) [inline forum/users/serializers.py:1]: Test the serializer to ensure the role is correctly added

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/users/urls.py [hunk 3]
```diff
@@ -1,7 +1,10 @@
 from django.urls import path
+from rest_framework_simplejwt.views import TokenObtainPairView, TokenRefreshView
+from .serializers import CustomTokenObtainPairSerializer
 
 app_name = 'users'
 
 urlpatterns = [
-    
+    path('sign-in/', TokenObtainPairView.as_view(), name='sign_in'),
+    path('token/refresh/', TokenRefreshView.as_view(), name='roken_refresh'),
```
Comment (devdsys) [inline forum/users/urls.py:9]: 1. Correct the name of the refresh path from 'roken_refresh' to 'token_refresh'.

2. It seems we already have an endpoint to refresh tokens provided by Djoser, like `/auth/jwt/refresh/` (refer to http://127.0.0.1:8000/swagger/). This might result in two endpoints with the same functionality. Please test both endpoints, choose the one that works best for you, and keep it.

File: forum/users/permissions.py
Comment (ne0ascorbinka) [reply forum/users/permissions.py:1]: I would agree if we used `User.role` field to determine his role; but in the meeting on Monday we agreed to use two-step authentication: first user signs in to his account and gets token with 'unassigned' role, and also gets two options to continue with - create/enter startup or investor profile. When entered existing profile, he is provided with token that contains his role 'investor' or 'startup'.
So it means we store his _current_ role in token, but roles _he can have in general_ in User model. 
Which means, that we should check his permissions by token, and as I know its body stored in `request.auth` as dictionary, which is why I used that method. I could look up to more robust way to check that permission tho

File: forum/users/serializers.py (hunk 1 above)
Comment (ne0ascorbinka) [reply forum/users/serializers.py]: Done. Now get_token uses enums; but again - I assume that due to authentication flow the responsibility to give certain role (investor/startup) when he enters profile is on the profile endpoints views.

File: forum/users/urls.py (hunk 3 above)
Comment (ne0ascorbinka) [reply forum/users/urls.py]: You're right, I fixe typo. Although, about endpoint - I think it's nice to keep them together, and maybe we don't really need all endpoints that djoser provides (all beside JWTs), at least rn

File: forum/users/permissions.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+from .models import Role, CustomUser
+
+class HasRole(BasePermission):
+    """
+    Permission class to check if a user has the required role,
+    based on both the token and the database state.
+    """
+
+    def __init__(self, role: Role):
+        self.role = role
```
Comment (mehalyna) [inline forum/users/permissions.py:12]: Permission classes should generally be stateless. Defining `__init__` with a dynamic parameter like `role` makes the permission class less reusable and deviates from standard DRF patterns.
DRF permissions are instantiated once and used across multiple requests. Using `__init__` dynamically can lead to unintended side effects.

File: forum/users/permissions.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    def has_permission(self, request: Request, view) -> bool:
+        if not request.user.is_authenticated:
+            return False
+
+        try:
+            token_role = request.auth.get('role')
```
Comment (mehalyna) [inline forum/users/permissions.py:19]: It’s better to explicitly verify the presence of `request.auth` and its structure before accessing it

File: forum/users/permissions.py
Comment (mehalyna) [inline forum/users/permissions.py:1]: The alignment logic between the token role and database role can be simplified and clarified.
Instead of dynamically passing the role to the permission class, create specific subclasses for each role. This aligns with DRF’s stateless permission design.
```
class IsUnassigned(HasRole):
    required_role = Role.UNASSIGNED


class IsInvestor(HasRole):
    required_role = Role.INVESTOR


class IsStartup(HasRole):
    required_role = Role.STARTUP
```

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/users/permissions.py
Comment (mehalyna) [reply forum/users/permissions.py:1]: Thank you for the detailed explanation regarding the use of `request.auth` for session-specific roles, especially in the context of your two-step authentication process. It makes sense to rely on the token for role-based permissions in this scenario since the role may evolve during the session, as users create or update profiles.
However, to ensure robustness, we could implement an additional layer of validation to gracefully handle edge cases where `request.auth` might not behave as expected (e.g., missing or malformed tokens). Perhaps a helper function or middleware could abstract this logic to make it more reusable and maintainable across the project.

File: forum/users/serializers.py (hunk 1 above)
Comment (mehalyna) [reply forum/users/serializers.py]: Given this, maintaining the role selection at the token level during login seems like the right approach. It allows the role to be explicitly set for the session without directly relying on the database field, which reflects the user's general roles rather than the session-specific one. This distinction is essential for the flow you’ve described.
One thing we could consider is ensuring a robust mechanism for validating the selected role during login. For example:
- Verify that the selected role aligns with one of the roles stored in the database for the user.
- Log the role-switching activity for better traceability.

File: forum/users/urls.py (hunk 3 above)
Comment (ne0ascorbinka) [reply forum/users/urls.py]: Also in latest commits I moved those routes to separate list, that is accessed with prefix 'auth/'

File: forum/users/permissions.py (hunk 5 above)
Comment (ne0ascorbinka) [reply forum/users/permissions.py]: Done

File: forum/users/serializers.py (hunk 2 above)
Comment (ne0ascorbinka) [reply forum/users/serializers.py]: Done: reworked serializer completely according to new one-step auth concept

File: forum/users/permissions.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            return False
+
+        if not isinstance(request.auth, AccessToken) or request.auth is None:
+            raise PermissionDenied("Something is wrong with auth: user's probably not using JWT or another problem")
+
+        token_role = request.auth.get('role')
+
+        if self.required_role == Role.UNASSIGNED:
+            return token_role == Role.UNASSIGNED.value
+
```
Comment (mehalyna) [inline forum/users/permissions.py:25]: Use dynamic checks for better scalability and maintainability. For example, encapsulate role validation logic within a helper method:
```
def is_role_valid(token_role, required_role):
    return token_role == required_role.value
```
Replace direct comparisons with calls to this function:
```
if self.required_role == Role.UNASSIGNED:
    return is_role_valid(token_role, Role.UNASSIGNED)
```

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/users/permissions.py (hunk 6 above)
Comment (ne0ascorbinka) [reply forum/users/permissions.py]: Done ✅

File: forum/users/serializers.py [hunk 7]
```diff
@@ -70,3 +71,31 @@ def validate_role(self, value):
         if value not in valid_roles:
             raise serializers.ValidationError("Invalid role.")
         return value
+
+
+VALID_TOKEN_ROLES = [Role.STARTUP, Role.INVESTOR]
+
+class CustomTokenObtainPairSerializer(TokenObtainPairSerializer):
```
Comment (devdsys) [inline forum/users/serializers.py:78]: The `role` field is required for authentication, but currently, only `email` and `password` are visible in Swagger by default. Please add the `role` field to the Swagger schema for the login endpoint.  

For example, you can add
```python
role = serializers.IntegerField(required=True)
```  
before 'validate' method at line 79.

Comment (ne0ascorbinka) [reply forum/users/serializers.py:78]: fr, done ✅

File: forum/users/permissions.py
Comment (mehalyna) [reply forum/users/permissions.py:1]: ?

Comment (mehalyna) [reply forum/users/permissions.py:1]: ?

Comment (mehalyna) [reply forum/users/permissions.py:1]: ?

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/users/permissions.py
Comment (ne0ascorbinka) [reply forum/users/permissions.py:1]: Done✅

File: forum/users/tests.py
Comment (mehalyna) [inline forum/users/tests.py:1]: Take into account few recommendations, pls:
- Include a test case for users with the `Role.UNASSIGNED` role to verify that they are correctly denied or granted access as per the permissions.
- Repeated code like setting up mock tokens (`self.mock_token.get = Mock(...)`) can be moved to a helper method to improve test readability and maintainability.
- Add a test to verify the behavior when a user attempts to request a token for a role they don't possess in the database.
- While there are tests for invalid authentication types, add a test specifically for scenarios where the request lacks an Authorization header.
- In tests where exceptions are raised (`ValidationError`, `PermissionDenied`), explicitly check the error message or relevant field to ensure clarity.
- Add a test case for when the `CustomTokenObtainPairSerializer` receives an empty request payload to ensure proper validation and error messaging.

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/users/tests/test_token_serializer.py
Comment (mehalyna) [inline forum/users/tests/test_token_serializer.py:1]: Consider adding tests for edge cases, such as invalid email formats or missing password fields.

   ```
   def test_invalid_email_format(self):
       """Test that an invalid email format raises a ValidationError."""
       request_data = {"email": "invalidemail", "password": "securepassword", "role": Role.INVESTOR.value}
       context = {"request": self._get_mock_request(request_data)}

       serializer = CustomTokenObtainPairSerializer(context=context)
       with self.assertRaises(ValidationError) as exc:
           serializer.validate(request_data)

       self.assertIn("email", exc.exception.detail)
       self.assertEqual(exc.exception.detail["email"], "Enter a valid email address.")
   ```

Comment (mehalyna) [inline forum/users/tests/test_token_serializer.py:1]: Instead of manually creating mocks, you can use `patch` from `unittest.mock` to temporarily replace objects during tests.

   ```
   from unittest.mock import patch

   @patch('users.serializers.CustomTokenObtainPairSerializer.validate')
   def test_missing_role(self, mock_validate):
       """Test that a missing role raises a ValidationError."""
       mock_validate.side_effect = ValidationError({"role": "This field is required."})
       request_data = {"email": self.user.email, "password": "securepassword"}  # No role provided
       context = {"request": self._get_mock_request(request_data)}

       serializer = CustomTokenObtainPairSerializer(context=context)
       with self.assertRaises(ValidationError) as exc:
           serializer.validate(request_data)

       self.assertIn("role", exc.exception.detail)
       self.assertEqual(exc.exception.detail["role"], "This field is required.")
   ```

**Review by mehalyna — APPROVED** [review]
Some recommendations and suggestions were added,  which I kindly ask you to take into account for the future

**Review by devdsys — APPROVED** [review]

**Review by danylkoltak — APPROVED** [review]
